### PR TITLE
[#1512] feature(Spark): Replace taskAttemptId with mapIndex in blockId

### PR DESCRIPTION
**Pull request by EnricoMi**
### What changes were proposed in this pull request?
Use map index (map task or source partition id) in block id instead of task attempt id.

### Why are the changes needed?
The map index has a lower cardinality than the unbound positive long (63 bit) task attempt id. This reduces the practical risk (fails with sufficient number of stages for 100TB dataset and 500k partitions) of overflowing sequence number, partition id or map index below an impractical level in today's terms (supports any number of stages for 1PT dataset with 16m partitions).

Fix: #1512

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Unit and integration tests.

### CODE REVIEW

File: integration-test/spark-common/src/test/java/org/apache/uniffle/test/FailingTasksTest.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import org.apache.spark.shuffle.RssSparkConfig;
+import org.apache.spark.sql.SparkSession;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import org.apache.uniffle.coordinator.CoordinatorConf;
+import org.apache.uniffle.server.ShuffleServerConf;
+import org.apache.uniffle.storage.util.StorageType;
+
+public class FailingTasksTest extends SparkIntegrationTestBase {
```
Comment (zuston) [inline integration-test/spark-common/src/test/java/org/apache/uniffle/test/FailingTasksTest.java:34]: This test case looks passed, which means the design looks effective. Please ping me if this is ready for review.

File: storage/src/main/java/org/apache/uniffle/storage/factory/ShuffleHandlerFactory.java [hunk 2]
```diff
@@ -126,6 +126,8 @@ private ClientReadHandler getMemoryClientReadHandler(
       Roaring64NavigableMap realExceptBlockIds = RssUtils.cloneBitMap(request.getExpectBlockIds());
       realExceptBlockIds.xor(request.getProcessBlockIds());
       expectTaskIds = RssUtils.generateTaskIdBitMap(realExceptBlockIds, request.getIdHelper());
+    } else {
+      expectTaskIds = request.getExpectTaskIds();
```
Comment (EnricoMi) [inline storage/src/main/java/org/apache/uniffle/storage/factory/ShuffleHandlerFactory.java:130]: Does this harm legacy Spark / TEZ / MR read clients?

Comment (EnricoMi) [reply storage/src/main/java/org/apache/uniffle/storage/factory/ShuffleHandlerFactory.java:130]: With `serverInfoList.size() == 1`, the `else` branch is used for TEZ and MR client. I think expected task ids and those task ids in block ids differ structurally for those clients. See related code in #1518.

I think with setting `.expectedTaskIdsBitmapFilterEnable(true)` for those two clients would fix this.

**Comment by connorlwilkes** [comment]
👍

**Comment by jerqi** [comment]
One question: We can't judge the data is from the same task. If one map task A have two attempts taskAttempt 1, taskAttempt 2. TaskAttempt 1 can have two blocks, TaskAttempt 2 may have three blocks. Because Spark doesn't guarantee every taskAttempts have the same data. Because some transforms of Spark are indeterministic.

**Comment by EnricoMi** [comment]
@jerqi you are right, if both tasks report their shuffle results (block ids), we would not know how many blocks to read for our partition id and map index. For that to work we need to guarantee that only one task attempt succeeds in registering shuffle results (blockids). We then take that taskAttemptId and those blockIds for the partition id and map index.

This would require a two-phase commit to guard / coordinate the report shuffle result step, which would be a much bigger change: https://github.com/G-Research/incubator-uniffle/pull/4/files

There are only two situations, where this can happen, though: speculative task execution (which is not recommended for production) and (without speculative task execution) a task failure after shuffle results have been reported, which triggers another task attempt.

**Comment by jerqi** [comment]
> @jerqi you are right, if both tasks report their shuffle results (block ids), we would not know how many blocks to read for our partition id and map index. For that to work we need to guarantee that only one task attempt succeeds in registering shuffle results (blockids). We then take that taskAttemptId and those blockIds for the partition id and map index.
> 
> This would require a two-phase commit to guard / coordinate the report shuffle result step, which would be a much bigger change: https://github.com/G-Research/incubator-uniffle/pull/4/files
> 
> There are only two situations, where this can happen, though: speculative task execution (which is not recommended for production) and (without speculative task execution) a task failure after shuffle results have been reported, which triggers another task attempt.

Spark has already had a commit mechanism. If we set up a new two-phrase commit. Maybe there will be some consistent.

**Comment by EnricoMi** [comment]
@jerqi Do you mean the RSS Spark clients or Apache Spark? Can you point me to it? What do you suggest to do next?

**Comment by jerqi** [comment]
We can use TaskSchedulerImpl#nextTaskId to simplify the taskAttemptId. 
1. We use Java reflection to get `TaskSchedulerImpl#nextTaskId` from SparkContext.
2. We record the start taskAttemptId in the ShuffleHandle.
3. We use taskAttemptId minus start taskAttemptId.

**Comment by jerqi** [comment]
> @jerqi Do you mean the RSS Spark clients or Apache Spark? Can you point me to it? What do you suggest to do next?

I mean Apache Spark has own commit mechinanism.

**Comment by EnricoMi** [comment]
> We can use TaskSchedulerImpl#nextTaskId to simplify the taskAttemptId.
> 
> - We use Java reflection to get TaskSchedulerImpl#nextTaskId from SparkContext.
> - We record the start taskAttemptId in the ShuffleHandle.
> - We use taskAttemptId minus start taskAttemptId.

Offsetting the taskAttemptId would help to reduce the ids for long running Spark jobs, but the offset ids can practically still be arbitrary large, e.g. if there are many consecutive large (with many partitions) stages.

What do you think about using the task id and the attempt no (bit-appended)? That id is unique for a stage and much more constraint than the taskAttemptId or the offset taskAttemptId.

> > @jerqi Do you mean the RSS Spark clients or Apache Spark? Can you point me to it? What do you suggest to do next?
> 
> I mean Apache Spark has own commit mechinanism.

Do you refer to [ShuffleMapOutputWriter](https://spark.apache.org/docs/latest/api/java/org/apache/spark/shuffle/api/ShuffleMapOutputWriter.html)?

**Comment by zuston** [comment]
Sorry for the late reply, we are in vocation for Chinese new year.

#1529 looks fine, can we add a integration test to cover the case of speculative exection?

**Comment by jerqi** [comment]
> > We can use TaskSchedulerImpl#nextTaskId to simplify the taskAttemptId.
> > 
> > * We use Java reflection to get TaskSchedulerImpl#nextTaskId from SparkContext.
> > * We record the start taskAttemptId in the ShuffleHandle.
> > * We use taskAttemptId minus start taskAttemptId.
> 
> Offsetting the taskAttemptId would help to reduce the ids for long running Spark jobs, but the offset ids can practically still be arbitrary large, e.g. if there are many consecutive large (with many partitions) stages.
> 
> What do you think about using the task id and the attempt no (bit-appended)? That id is unique for a stage and much more constraint than the taskAttemptId or the offset taskAttemptId: #1529
> 
> > > @jerqi Do you mean the RSS Spark clients or Apache Spark? Can you point me to it? What do you suggest to do next?
> > 
> > 
> > I mean Apache Spark has own commit mechinanism.
> 
> Do you refer to [ShuffleMapOutputWriter](https://spark.apache.org/docs/latest/api/java/org/apache/spark/shuffle/api/ShuffleMapOutputWriter.html)?

AttemptNo will waste some bits. If we increase the bits, the bitmap will occupy more memory.

**Comment by EnricoMi** [comment]
> AttemptNo will waste some bits. If we increase the bits, the bitmap will occupy more memory.

See https://github.com/apache/incubator-uniffle/pull/1529#issuecomment-1947826787.

**Comment by EnricoMi** [comment]
> #1529 looks fine, can we add a integration test to cover the case of speculative exection?

See https://github.com/apache/incubator-uniffle/pull/1529#issuecomment-1947835597.

**Comment by EnricoMi** [comment]
Superseded by #1529..

**Pull request closed without merging**